packages/flutter/lib/src/rendering/object.dart

```diff
@@ -3248,30 +3249,87 @@ abstract class RenderObject with DiagnosticableTreeMixin implements HitTestTarge
   /// the global coordinate system in logical pixels. To get physical pixels,
   /// use [applyPaintTransform] from the [RenderView] to further transform the
   /// coordinate.
+  ///
+  /// See also:
+  ///
+  ///  * [computeTransformToRenderObject] which does the same, but the target
+  ///    `RenderObject` does not have to be an ancestor.
   Matrix4 getTransformTo(RenderObject? ancestor) {
-    final bool ancestorSpecified = ancestor != null;
     assert(attached);
-    if (ancestor == null) {
-      final RenderObject? rootNode = owner!.rootNode;
-      if (rootNode is RenderObject) {
-        ancestor = rootNode;
-      }
+    final Matrix4 identity = Matrix4.identity();
+    if (ancestor != null) {
+      return _computePaintTransform(ancestor, identity);
+    } else  {
+      final RenderObject? parent = this.parent;
+      return parent != null ? _computePaintTransformToGlobalCoordinates(parent, identity) : identity;
     }
-    final List<RenderObject> renderers = <RenderObject>[];
-    for (RenderObject renderer = this; renderer != ancestor; renderer = renderer.parent!) {
-      renderers.add(renderer);
-      assert(renderer.parent != null); // Failed to find ancestor in parent chain.
+  }
+
+  // Computes the paint transform from this to toRenderObject and applies it on
+  // `transform`.
+  Matrix4 _computePaintTransform(RenderObject toRenderObject, Matrix4 transform) {
+    if (!identical(this, toRenderObject)) {
+      final RenderObject? parent = this.parent;
+      assert(
+        toRenderObject.depth <= depth && parent != null,
+        '$toRenderObject is not an ancestor of the given RenderObject. Consider using computeTransformToRenderObject instead.',
+      );
+      parent!.applyPaintTransform(this, parent._computePaintTransform(toRenderObject, transform));
     }
-    if (ancestorSpecified) {
-      renderers.add(ancestor!);
+    return transform;
+  }
+
+  Matrix4 _computePaintTransformToGlobalCoordinates(RenderObject parent, Matrix4 transform) {
+    final RenderObject? parentParent = parent.parent;
+    if (parentParent != null) {
+      parent.applyPaintTransform(this, parent._computePaintTransformToGlobalCoordinates(parentParent, transform));
     }
+    assert(parentParent != null || parent is RenderView);
+    assert(parentParent == null || parent is! RenderView);
+    return transform;
+  }
+
+  /// Returns the paint transform that allows this `RenderObject` to perform
+  /// paint in `toRenderObject`'s coordinate system.
+  ///
+  /// Returns null if `toRenderObject` is not in the same render tree, or either
+  /// of the two `RenderObject`s is in an offscreen subtree.
+  ///
+  /// See also:
+  ///
+  ///  * [getTransformTo] which does the same, but only searches up the tree.
+  ///  * [paintsChild] which this method uses to determine whether
+  ///    a `RenderObject` is in an offscreen subtree.
+  Matrix4? computeTransformToRenderObject(RenderObject toRenderObject) {
```

Comment:
I was a bit concerned that someone would actually depend on the current behavior of `getTransformTo` but since it just asserts I guess that's unlikely. I'll change the behavior of `getTransformTo` instead.